docs(design): Scan Container Images with Trivy Filesystem Scanne (#830)
aquasecurity/starboard: docs/design/design_trivy_file_system_scanner.md

3. Vendor-specific or local extension. For example, methods described on [AWS ECR Private registry authentication].

Even though we could resolve some of above-mentioned limitations with hostPath volume mounts to the container runtime
socket, it would have its own disadvantages that we are trying to avoid. For example, more permissions to schedule scan
Jobs and additional information about cluster's infrastructure such as location of the container runtime socket. 

## Solution

### TL;DR;

Use Trivy filesystem scanning to scan container images. The main idea, which is discussed in this proposal, is to
schedule a scan Job on the same cluster node where the scanned workload. This allows Trivy to scan a filesystem of
the container image which is already cached on that node without pulling the image from a remote registry. What's more,
Trivy will scan container images from private registries without providing registry credentials (as ImagePullSecret or
in any other proprietary way).

### Deep Dive

To scan a container image of a given K8s workload Starboard will create a corresponding container of a scan Job and
override its entrypoint to invoke Trivy filesystem scanner.

> This approach requires Trivy executable to be downloaded and made available to the entrypoint. We'll do that by adding

Reviewer: This could maybe be done with a csi driver as well.

Reviewer 2: I'm not sure that I understood this comment. Could you elaborate on how we can use csi driver in this case?

Reviewer: just thinking, instead of copying the file from an init container to an emtpydir, a csi driver in ephemeral mode could mount the file into the container.

Reviewer: May not be worth the effort. but could be like:

kind: Pod
apiVersion: v1
metadata:
  name: my-csi-app-inline-volume
spec:
  containers:
    - name: my-frontend
      image: busybox
      command: [ "sleep", "100000" ]
      volumeMounts:
      - mountPath: "/trivy"
        name: my-csi-volume
  volumes:
  - name: my-csi-volume
    csi:
      driver: trivy

the init container to the scan Job. Such init container will use the Trivy container image to copy Trivy executable out
to the emptyDir volume, which will be shared with the other containers.

> Another init container is required to download Trivy vulnerability database and save it to the mounted shared volume.

Reviewer: Same.

Finally, the scan container will use shared volume with the Trivy executable and Trivy database to perform the actual
filesystem scan. (See the provided [Example](#example) to have a better idea of all containers defined by a scan Job and
how they share data via the emptyDir volume.)

> Note that the second init container is required in [Standalone] mode, which is the only mode supported by Trivy
> filesystem scanner at the time of writing this proposal.

We further restrict scan Jobs to run on the same node where scanned Pod is running and never pull images from remote
registries by setting the `ImagePullPolicy` to `Never`. To determine the node for a scan Job Starboard will list active
Pods controlled by the scanned workload. If the list is not empty it will take the node name from the first Pod,

> otherwise it will ignore the workload.

Reviewer: This may run afoul of this admission controller https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#alwayspullimages

Reviewer 2: That's a very good point. We've merge this PR already but I opened #900 to add you comments as remark.

Author(s): [ [Devendra Turkar](https://github.com/deven0t), [Daniel Pacak](https://github.com/danielpacak) ]

## Abstract

Add additional support in Trivy plugin to support file system scanning of an image. 

## Background

Starboard currently uses Trivy  plugin to scan and generate Vulnerability Report for an image. Trivy can scan a container 
image by pulling the image from a remote registry. This mechanism works good for public registry image, but for private 
registry image Trivy is dependent on registry credentials. Currently, starboard has mechanism to provide these credentials
using ImagePullSecret or with ServiceAccount, based on how it is provided to application pod.
With this approach we have following problem:
1. When image pulled from private registry without ImagePullSecret or service account 
(https://kubernetes.io/docs/concepts/containers/images/#configuring-nodes-to-authenticate-to-a-private-registry)  
2. When image pulled from managed registry of managed cluster
3. When pod is running with image present on host

## Proposal

Use Trivy file system scanning option to scan an image. So that we will not need to provide registry credentials to trivy.

> K8s scan job will use underline secrets available to pull an image and that image will be scanned by trivy as file system

Reviewer 2: For the current proposal the main idea is to schedule a scan job on the same node where the scanned workload so we can leverage container image cached on that node and scan without providing credentials to Trivy.

```shell script
trivy fs --ignore-unfixed /
```
With above command trivy will scan entire file system of an image and it will give us the result.

To run trivy in that image we need trivy to be available in the container. To solve this problem we have a way for it.
1. Add an init container with tvivy image in the scan job which will pull trivy image, Copy trivy executable in a volume
 which will be shared with main container 
3. Add another init container to download trivy database and mount db file in volume which will be used as trivy database
during file system scan

### TL;DR;

Currently, clientServer mode is not supporting filesystem scan. So we will have to go with standalone mode. In future we
 can change it to client server mode and accordingly we will have to change our scan job spec.
 
 Configuration parameter can be added to enable this scan mode
``trivy.command: fs``. This trivy.command will support two values ``image`` and ``fs``. For backward compatibility. if 
this config parameter is not defined then it will consider it as ``image``.

We have two options to consider here
1. We can restrict scan job to run on same host where original pod is running.

>     - ImagePullPolicy can be set as false, so we will not need to pull the image on node

Reviewer 2: ```suggestion
    - ImagePullPolicy will be set to `Never`, so we will not need to pull the image on node
```

    - nodeName has to be provided for scan job to run on specific node
2. Allow scan job to run on any node in the cluster.  *This option will not be implemented for now*
    - ImagePullPolicy should be set as IfNotPresent
    - nodeName will be set as empty, so that scan job will run on any node.
    - make ImagePullSecret available for scan job
    
## Scan job running on same host

To run scan on same host where application pod is running then we have to identify host on which application pod is 
running. To do this we will search for active pod for any controller and we will get the nodename from first pod of it.

1. Scan job has to run on same host where pod is running.
    - Get the nodename of one of the pod. If we dont get nodename then return error.
    - No need to copy secret from the source namespace to starboard-operator namespace. We will ignore imagePullSecret of 
    source application workload.
    - ImagePullPolicy will be set as Never

>     - nodeName will be passed explicitly.

Reviewer 2: Clarify that we set nodeName on a scan job so it's scheduled on the same node as scanned workload.

          volumeMounts:
            - name: scan-volume
              mountPath: /var/starboard
         # The trivy-download-db container is using trivy executable binary
         # from the previous step to download Trivy vulnerability database
         # from GitHub releases page.
        - name: trivy-download-db
          image: aquasec/trivy:0.19.2
          command:
            - /var/starboard/trivy
            - --download-db-only
            - --cache-dir
            - /var/starboard/trivy-db
          volumeMounts:
            - name: scan-volume
              mountPath: /var/starboard
      containers:
        # The scan-nginx container is based on the container image that
        # we want to scan with Trivy. However, it has overwritten command (entrypoint)
        # to invoke trivy file system scan. The scan results are output to stdout
        # in JSON format so we can parse them and store as VulnerabiltyReport instance.
        - name: scan-nginx
          image: example.registry.com/nginx:1.16
          # To scan image layers cached on K8s node without pulling
          # it from a remote registry.
          imagePullPolicy: Never

>           command:

Reviewer 2: We are changing the entrypoint here, but I think we should also mention that Trivy requires root permissions. IIRC a scan Job must run as UID 0. Does it have any impact on PSP or any solution that supersedes PSPs?

Author: Yes, we need to run this scan job container with root user `runAsUser: 0`. This will have impact with PSP MustRunAsNonRoot

            - /var/starboard/trivy
            - --cache-dir
            - /var/starboard/trivy-db
            - fs
            - --format
            - json
            - --ignore-unfixed
            - /
          volumeMounts:
            - name: scan-volume
              mountPath: /var/starboard
```

##### Current Limitations for this option:
1. If controller is created, and it doesn’t have any active pods then this mode will not scan it

> 2. We cannot identify pods running for cronjob, so those will not be scanned.

Reviewer 2: Can you elaborate on why we cannot identify pods running for CronJobs? Is it because we scan by K8s controller not by pod?

Author: yes, boz when starboard receives event of cronjob we may not have an active pod running for it. and also getting active pods of cron job is not straight forward way, so added it as limitation for now.
this is one way to get jobs for cronjob, later we will have to get active pods for these jobs. https://github.com/kubernetes/kubernetes/blob/e640a01219736dd051dc3a9b8ac4630ec18b6b25/pkg/controller/cronjob/cronjob_controllerv2.go#L226
